pkg/operator/controllers/dnsmasq/cluster_controller.go

```diff
@@ -0,0 +1,104 @@
+package dnsmasq
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+
+	"github.com/openshift/installer/pkg/aro/dnsmasq"
+	mcoclient "github.com/openshift/machine-config-operator/pkg/generated/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	aroclient "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers"
+	"github.com/Azure/ARO-RP/pkg/util/dynamichelper"
+)
+
+type ClusterReconciler struct {
+	log *logrus.Entry
+
+	arocli aroclient.Interface
+	mcocli mcoclient.Interface
+	dh     dynamichelper.Interface
+}
+
+func NewClusterReconciler(log *logrus.Entry, arocli aroclient.Interface, mcocli mcoclient.Interface, dh dynamichelper.Interface) *ClusterReconciler {
+	return &ClusterReconciler{
+		log:    log,
+		arocli: arocli,
+		mcocli: mcocli,
+		dh:     dh,
+	}
+}
+
+// Reconcile watches the ARO object, and if it changes, reconciles all the
+// 99-%s-aro-dns machineconfigs
+func (r *ClusterReconciler) Reconcile(request ctrl.Request) (ctrl.Result, error) {
+	// TODO(mj): controller-runtime master fixes the need for this (https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/reconcile/reconcile.go#L93) but it's not yet released.
+	ctx := context.Background()
+	if request.Name != arov1alpha1.SingletonClusterName {
+		return reconcile.Result{}, nil
+	}
+
+	mcps, err := r.mcocli.MachineconfigurationV1().MachineConfigPools().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		r.log.Error(err)
+		return reconcile.Result{}, err
+	}
+
+	roles := make([]string, 0, len(mcps.Items))
+	for _, mcp := range mcps.Items {
+		roles = append(roles, mcp.Name)
+	}
+
+	err = reconcileRoles(ctx, r.arocli, r.dh, roles...)
```

Comment:
`reconcileRole` name is very confusing. What roles this is? Kubernetes roles? Node roles? Azure roles? Took me a while to read down the rabbit hole to understand.

Reply:
Agree, will fix and document